lib/_emerge/actions.py

```diff
@@ -842,7 +842,7 @@ def action_depclean(
                 )
         if not matched_packages:
             writemsg_level(f">>> No packages selected for removal by {action}\n")
-            return 0
+            return 1
```

Comment:
What does this give with bare `emerge -vc` if there's nothing for it to do?

Reply:
```
~/gentoo/portage $ sudo ./bin/emerge -vc; echo $?
[copious output]
  x11-wm/i3-4.22 pulled in by:
    x11-misc/polybar-3.6.3-r3 requires x11-wm/i3

>>> No packages selected for removal by depclean
Packages installed:   2356
Packages in world:    431
Packages in system:   49
Required packages:    2356
Number removed:       0
0
~/gentoo/portage $ git log -1
commit 3edfa9ced9cb0455239f6cc47867f1ebd9bcb203 (HEAD -> bug917120, fork/bug917120)
Author: John Helmert III <ajak@gentoo.org>
Date:   Wed Nov 15 19:45:14 2023 -0800

    depclean: return failure when no packages are selected for depclean
```

The code in question is:

```
    if myfiles:
        args_set.update(myfiles)
        matched_packages = False
        for x in args_set:
            if vardb.match(x):
                matched_packages = True
            else:
                writemsg_level(
                    f"--- Couldn't find '{x.replace('null/', '')}' to {action}.\n",
                    level=logging.WARN,
                    noiselevel=-1,
                )
        if not matched_packages:
            writemsg_level(f">>> No packages selected for removal by {action}\n")
            return 1

    # The calculation is done in a separate function so that depgraph
    # references go out of scope and the corresponding memory
    # is freed before we call unmerge().
    rval, cleanlist, ordered, req_pkg_count = calc_depclean(
        settings, trees, ldpath_mtimes, myopts, action, args_set, spinner
    )
```

So, it looks like it tries to resolve the packages specified on the command line first, and then returns if no packages were matched based on the command line arguments, but it doesn't follow that path at all if there's nothing on the command line. In that case, execution flows right into the below `calc_depclean` which does exit with success if there's nothing to do.